compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2785,11 +2785,9 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
       checkDerivedValueClass(cls, body1)
 
       val effectiveOwner = cls.owner.skipWeakOwner
-      if !cls.isRefinementClass
-         && !cls.isAllOf(PrivateLocal)
+      if cls.is(ModuleClass)
```

Comment:
We do not have any `neg` tests for this case. The following seems to cover what we are missing.

```scala
trait Outer extends Any {
  trait Inner1
  trait Inner2 extends Any
  class Inner3
  class Inner4(a: Int) extends AnyVal // error
  case class Inner5(a: Int) // error
  object Inner6 // error
}
```